src/Datastore/RAMDataStore.cpp

```diff
@@ -0,0 +1,337 @@
+      // duplicate, do nothing
+    } else {
+      // Track already in vote vector and we want to remove the upvote:
+      // we want to remove it from upvoted tracks, so remove it from vector of
+      // upvoted tracks and update vote counter in track
+      it->votes.erase(it_track);
+      // decrement its upvote counter
+      if (pAdminTrack != nullptr) {
+        pAdminTrack->votes--;
+      }
+      if (pNormalTrack != nullptr) {
+        pNormalTrack->votes--;
+      }
+    }
+  } else {
+    // Track not in vote vector
+    if (vote) {
+      // Track not in vote vector and we want to upvote it: add to vector and
+      // update counter
+      it->votes.emplace_back(tID);
+      // increment its upvote counter
+      if (pAdminTrack != nullptr) {
+        pAdminTrack->votes++;
+      }
+      if (pNormalTrack != nullptr) {
+        pNormalTrack->votes++;
+      }
+    } else {
+      // track not in vote vector and we want to remove upvote: cant remove
+      // nonexistent upvote, so do nothing
+    }
+  }
+
+  // sort Normal Queue
+  sort(mNormalQueue.tracks.begin(), mNormalQueue.tracks.end());
+  return nullopt;
+}
+
+TResult<Queue> RAMDataStore::getQueue(QueueType q) {
+  // Shared Access to Song Queue
+  shared_lock<shared_mutex> MyLock(mQueueMutex, defer_lock);
+  MyLock.lock();
+
+  // select Queue
+  Queue *pQueue = SelectQueue(q);
+  if (pQueue == nullptr) {
+    return Error(ErrorCode::InvalidValue, "Invalid Parameter in Queue");
+  }
+
+  // return read only access to Queue
+  return (const Queue)(*pQueue);
+}
+
+TResult<PlaybackTrack> RAMDataStore::getPlayingTrack() {
+  // Shared Access to Song Queue
+  shared_lock<shared_mutex> MyLock(mQueueMutex, defer_lock);
+  MyLock.lock();
+
+  return mCurrentTrack;
+}
+
+TResult<bool> RAMDataStore::hasUser(TSessionID ID) {
+  // Shared Access to User List
+  shared_lock<shared_mutex> MyLock(mUserMutex, defer_lock);
+  MyLock.lock();
+
+  // find user
+  User user;
+  user.SessionID = ID;
+  auto it = find(mUsers.begin(), mUsers.end(), user);
+  if (it == mUsers.end()) {
+    return false;
+  } else {
+    return true;
+  }
+}
+
+TResultOpt RAMDataStore::nextTrack() {
+  // Exclusive Access to Song Queue
+  unique_lock<shared_mutex> MyLock(mQueueMutex, defer_lock);
+  MyLock.lock();
+
+  // Increment LastPlayed counter for all songs
+
+  // doesnt work with auto for some reason
+  //  for(auto elem : mAdminQueue.tracks){
+  //      elem.LastPlayedxSongsAgo += 1;
+  //  }
+  //    for(auto elem : mNormalQueue.tracks){
+  //        elem.LastPlayedxSongsAgo += 1;
+  //    }
+  for (int i = 0; i < mAdminQueue.tracks.size(); ++i) {
+    mAdminQueue.tracks[i].LastPlayedxSongsAgo += 1;
+  }
+  for (int i = 0; i < mNormalQueue.tracks.size(); ++i) {
+    mNormalQueue.tracks[i].LastPlayedxSongsAgo += 1;
+  }
+
+  QueuedTrack tr;
+
+  // If there are songs in the Admin Queue, play the first of those and move it
+  // to the user queue
+  if (mAdminQueue.tracks.size()) {
```

Comment:
why do you want the current track in user queue at all? that's what `mCurrentTrack` is for

Reply:
I assumed that tracks should be removed from the admin queue after being played once because otherwise we`ll never get around to the user queue. I also assumed that if a track is added to the admin queue but not the user queue that it should be played once from admin and then moved to user

Reply:
there is no moving involved at all. the distinction between admin and user queue is only to determine the order in which tracks get played.

as long as tracks are in the admin queue -> play the next song from that queue
if there are no tracks in the admin queue -> play the next song from the user queue

when a track is played it is removed from the queue and stored in `mCurrentTrack`, because it is the current track now 

once a track finished playing it no longer is in any queue nor in `mCurrentTrack`

Reply:
where did we specify that tracks should be deleted after playing? i cant find that in either of our Requirements Documents:
https://docs.google.com/document/d/1qKVGvLo-GKIQYtxzhYwbMTYPCDQOG4stTbK1WtuHjFA/edit#heading=h.r8a9yfz7tbmn
or in 
https://docs.google.com/document/d/1U1Do0q0Cut-wy9hFlZ2SG6rTCJi91WWktYxBhr_Th6s/edit

Reply:
it seems like it is not explicitly defined anywhere, but neither is the other case.

Reply:
well if its not defined and works why change it? I wrote a perfectly good scheduler which doesnt violate the specs and i dont want to throw it away. Especially since throwing it away would mean that i would have to adapt the test cases as well

Reply:
normally if something isnt explicitly specified you decide as a team what to do (especially such drastic changes).
lets see what @wurmmi  and @janeson332 think about it

Reply:
i would also go for the option of simply removing tracks, once they were played.

but let's find 2 minutes tomorrow and discuss meaningful arguments about that matter.